### DISCUSSION TITLE
Cannot A/B test static pages using middleware

### DISCUSSION THREAD

**Discussion by zmarty** [vercel/next.js#33670] [Help]
### What version of Next.js are you using?

12.0.7

### What version of Node.js are you using?

16.13.1

### What browser are you using?

Edge (Chromium)

### What operating system are you using?

Windows 10

### How are you deploying your application?

Azure App Service

### Describe the Bug

Relevant information specifically for this bug: We host Next.js in Azure on our own in Azure.

We tried to run an A/B experiment using the new middleware functionality on two **static** pages: index.tsx and index2.tsx.

The problem we are having is that for static pages Next.js emits a cache control header that makes our CDN cache one page or another. And we cannot override cache control headers (to set it to no-cache) for static pages in next.config.js. Here's what the [docs](https://nextjs.org/docs/going-to-production#caching) say: Cache-Control headers set in next.config.js will be overwritten in production to ensure that static assets can be cached effectively.

### Expected Behavior

- Do not emit cache control headers that cache a page when it is served through a rewrite from middleware.
- OR: Allow overriding Cache-Control header for static pages in next.config.js, at least if middleware is involved in rewriting them.

### To Reproduce

Create two static pages, have a _middleware randomly choose between them. Notice if you put a CDN in front of the website that A/B testing will fail to work as expected because one of the sides (A or B) will get cached by the CDN.

**Comment by balazsorban44** [comment]
This sounds like a platform-specific issue that you would have to take up with Azure.

We have proven/working A/B testing examples with middleware, you likely have seen it: https://github.com/vercel/examples/tree/main/edge-functions%2Fab-testing-simple

**Reply by zmarty** [reply]
I am sorry I mentioned Azure at all because this happens outside of Azure too as long as Next.js is self-hosted. Can you please re-read the description and give your opinion?

I am aware of that example and that works fine **as long as the pages are not static**.

**Reply by zmarty** [reply]
I created a repro here using an empty Next.js 12.0.8 (yarn create next-app) to demonstrate: 
https://github.com/zmarty/nextjs-middleware-static-bug/tree/main/pages/repro

Please start it with yarn build and yarn start, not with yarn dev.

Here is the header I get when I visit http://localhost:3000/repro

![image](https://user-images.githubusercontent.com/330396/151116068-d00ce18f-bd2a-4b1c-b4c7-7e57a16bc006.png)

If I place a CDN in front of this self-hosted Next.js instance, the A/B testing will fail since either A or B will get cached by the CDN.

I am also pasting the code here, if it disappears later:

### _middleware.ts

```typescript
import { NextRequest, NextResponse } from 'next/server'

export function middleware(req: NextRequest) {
  return Math.random() >= 0.5 ? NextResponse.rewrite(`/repro/pageA`) :  NextResponse.rewrite(`/repro/pageB`);
}
```

### pageA.js

```typescript
import Head from 'next/head'
import styles from '../../styles/Home.module.css'

export async function getStaticProps(context) {
  return {
    props: {}, // will be passed to the page component as props
  }
}

export default function Home() {
  return (
    <div className={styles.container}>
      <Head>
        <title>Page A</title>
      </Head>

      <main className={styles.main}>
        <h1 className={styles.title}>
          Page A
        </h1>
      </main>
    </div>
  )
}
```

### pageB.js

```typescript
import Head from 'next/head'
import styles from '../../styles/Home.module.css'

export async function getStaticProps(context) {
  return {
    props: {}, // will be passed to the page component as props
  }
}

export default function Home() {
  return (
    <div className={styles.container}>
      <Head>
        <title>Page B</title>
      </Head>

      <main className={styles.main}>
        <h1 className={styles.title}>
          Page B
        </h1>
      </main>
    </div>
  )
}
```

**Reply by zmarty** [reply]
@balazsorban44  Please take a look and if you agree, please switch this back to a bug.

**Reply by balazsorban44** [reply]
Thank you, we will look more into this.

**Reply by zmarty** [reply]
@balazsorban44 We are happy to look into this ourselves but I don't have a complete mental model of why this works on Vercel and it does not work in self-hosted mode. I don't entirely understand how/if your Cloudflare workers run the middleware separately from the pages themselves? and how that interacts with your CDN cache. Maybe some code pointers?

**Reply by lukebussey** [reply]
You need to use the Vary header to indicate to the cache (any CDN) that there are multiple variations of the page.

**Reply by zmarty** [reply]
Sorry but I don't see how that would help. The data flow is like this Azure FrontDoor CDN -> Our Frontend. If the CDN already cached the page, how would A/B testing work?

**Reply by lukebussey** [reply]
Every CDN has a cache key. Usually it'll be the full URL including query string, but if you serve different content for the same full URL as in the case with a server side a/b test, then you'll need to tell the CDN which headers to also include in the cache key.

A good example of this is when your origin server performs content negotiation to determine if the client can accept gzipped responses. In this case there will be two objects stored in cache for the same URL. One uncompressed, and one compressed. In this case the CDN is using the Content-encoding header as part of it's cache key.

- https://developers.cloudflare.com/cache/about/cache-keys
- https://www.fastly.com/blog/ab-testing-edge#keeping-different-versions-cached-with-http-vary

**Reply by stigkj** [reply]
How would one do an A/B test using the `Vary` header? I mean, you would need to have something that is sent from the browser on each request and can be unique for a group of users. As far as I can see, only a cookie can do this, and using `Vary: Cookie` does not really work. You can of course add some logic to your CDN to read from the cookie and use that to create a cache key.

**Reply by zmarty** [reply]
Yeah but what would set the cookie in the first place? Anyway I don't think using Vary header would work.

Our ask is this: If we use middleware to switch between two static pages, then let us override the cache header in next.config.js instead of forcing it like it does for all static pages. So then we would make this route return no-cache, middleware will run on every request, and CDN will not cache this route. Then A/B testing will work. Then when A/B test is done, we will remove the middleware, make one of the two static pages be the official one, and re-enable the caching header.

**Reply by lukebussey** [reply]
Yes you could prevent the a/b tested page from caching but the `no-cache` Cache-control directive doesn't prevent caching. It tells the browser and cdn to revalidate.

`Cache-control: private, no-store` will prevent caching on a CDN

Alternatively if the CDN obeys the Vary header, then you could try something like this in the middleware:

```
X-Variation: <variation name>
Vary: X-Variation
```

**Reply by stigkj** [reply]
Ok...now I'm just wondering how the CDN knows how to choose A or B when a client accesses the page? I.e. in A/B tests, it is often necessary that users always get one of the versions.

**Reply by lukebussey** [reply]
If the backend sets a cookie the CDN shouldn't cache the response.
